# Bellabeat Case Study
## How Can a Wellness Technology Company Play It Smart?

**Tools Used:** Python, SQL (PostgreSQL), Plotly  
**Data Source:** FitBit Fitness Tracker Data (Kaggle)

### Background
Bellabeat is a high-tech company that manufactures health-focused
smart products for women. This analysis explores smart device usage
data to gain insight into how consumers use non-Bellabeat smart
devices, and apply these insights to guide Bellabeat's marketing strategy.

### Business Task
Analyze FitBit fitness tracker data to identify trends in smart device
usage and provide data-driven recommendations for Bellabeat's marketing team.

In [67]:
import pandas as pd
import numpy as np
import os
!pip install plotly
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio

pio.renderers.default = 'iframe'

In [68]:
import kagglehub
path = kagglehub.dataset_download("sawandikirby/bellabeat-case-study-fitbit-data")

Using Colab cache for faster access to the 'bellabeat-case-study-fitbit-data' dataset.


In [69]:
import os
for root, dirs, files in os.walk(path):
  for file in files:
    if file.endswith('.csv'):
      print(os.path.join(root, file))

/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/minuteCaloriesNarrow_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/weightLogInfo_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/sleepDay_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/dailyIntensities_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/minuteIntensitiesWide_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/minuteMETsNarrow_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/dailyCalories_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/hourlyCalories_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 4.12.16-5.12.16/heartrate_seconds_merged.csv
/kaggle/input/bellabeat-case-study-fitbit-data/Fitabase Data 

In [70]:
data_dir = path + "/Fitabase Data 4.12.16-5.12.16"
daily_activity_path = data_dir + "/dailyActivity_merged.csv"
sleep_path = data_dir + "/sleepDay_merged.csv"
hourly_steps_path = data_dir + "/hourlySteps_merged.csv"
hourly_calories_path = data_dir + "/hourlyCalories_merged.csv"
hourly_intensities_path = data_dir + "/hourlyIntensities_merged.csv"
heartrate_path = data_dir + "/heartrate_seconds_merged.csv"

In [71]:
daily_activity = pd.read_csv(daily_activity_path)
daily_activity.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,4/12/2016,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,4/13/2016,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,4/14/2016,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,4/15/2016,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,4/16/2016,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


In [72]:
daily_activity.shape

(940, 15)

In [73]:
daily_activity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        940 non-null    int64  
 1   ActivityDate              940 non-null    object 
 2   TotalSteps                940 non-null    int64  
 3   TotalDistance             940 non-null    float64
 4   TrackerDistance           940 non-null    float64
 5   LoggedActivitiesDistance  940 non-null    float64
 6   VeryActiveDistance        940 non-null    float64
 7   ModeratelyActiveDistance  940 non-null    float64
 8   LightActiveDistance       940 non-null    float64
 9   SedentaryActiveDistance   940 non-null    float64
 10  VeryActiveMinutes         940 non-null    int64  
 11  FairlyActiveMinutes       940 non-null    int64  
 12  LightlyActiveMinutes      940 non-null    int64  
 13  SedentaryMinutes          940 non-null    int64  
 14  Calories  

## 1. Data Preparation

### Data Source
- **Dataset:** FitBit Fitness Tracker Data (Kaggle)
- **Period:** April 12, 2016 – May 12, 2016
- **Sample Size:** 33 FitBit users

### Files Used
| File | Description |
|------|-------------|
| dailyActivity_merged.csv | Daily steps, calories, active minutes |
| sleepDay_merged.csv | Sleep duration and quality |
| hourlySteps/Calories/Intensities | Hourly activity breakdown |
| heartrate_seconds_merged.csv | Heart rate (downsampled to hourly) |

### Cleaning Steps
- Converted `Id` to string format
- Converted date columns to datetime
- Removed 3 duplicate records from sleep data
- Downsampled heart rate from seconds to hourly averages
- Merged 3 hourly files into single table

In [74]:
daily_activity['Id'] = daily_activity['Id'].astype(str)
daily_activity['ActivityDate'] = pd.to_datetime(daily_activity['ActivityDate'])

In [75]:
daily_activity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        940 non-null    object        
 1   ActivityDate              940 non-null    datetime64[ns]
 2   TotalSteps                940 non-null    int64         
 3   TotalDistance             940 non-null    float64       
 4   TrackerDistance           940 non-null    float64       
 5   LoggedActivitiesDistance  940 non-null    float64       
 6   VeryActiveDistance        940 non-null    float64       
 7   ModeratelyActiveDistance  940 non-null    float64       
 8   LightActiveDistance       940 non-null    float64       
 9   SedentaryActiveDistance   940 non-null    float64       
 10  VeryActiveMinutes         940 non-null    int64         
 11  FairlyActiveMinutes       940 non-null    int64         
 12  LightlyActiveMinutes  

In [76]:
daily_activity.duplicated().sum()

np.int64(0)

In [77]:
daily_activity.columns = daily_activity.columns.str.lower()

In [78]:
daily_activity.columns

Index(['id', 'activitydate', 'totalsteps', 'totaldistance', 'trackerdistance',
       'loggedactivitiesdistance', 'veryactivedistance',
       'moderatelyactivedistance', 'lightactivedistance',
       'sedentaryactivedistance', 'veryactiveminutes', 'fairlyactiveminutes',
       'lightlyactiveminutes', 'sedentaryminutes', 'calories'],
      dtype='object')

In [79]:
def clean_daily_activity(path):
    df = pd.read_csv(path)
    df['Id'] = df['Id'].astype(str)
    df['ActivityDate'] = pd.to_datetime(df['ActivityDate'])
    df.columns = df.columns.str.lower()
    return df

In [80]:
daily_activity = clean_daily_activity(daily_activity_path)
daily_activity.head()

,id,activitydate,totalsteps,totaldistance,trackerdistance,loggedactivitiesdistance,veryactivedistance,moderatelyactivedistance,lightactivedistance,sedentaryactivedistance,veryactiveminutes,fairlyactiveminutes,lightlyactiveminutes,sedentaryminutes,calories
0,1503960366,2016-04-12,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,2016-04-13,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,2016-04-14,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,2016-04-15,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,2016-04-16,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


In [81]:
sleep_raw = pd.read_csv(sleep_path)
sleep_raw.head()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
0,1503960366,4/12/2016 12:00:00 AM,1,327,346
1,1503960366,4/13/2016 12:00:00 AM,2,384,407
2,1503960366,4/15/2016 12:00:00 AM,1,412,442
3,1503960366,4/16/2016 12:00:00 AM,2,340,367
4,1503960366,4/17/2016 12:00:00 AM,1,700,712


In [82]:
sleep_raw.duplicated().sum()

np.int64(3)

In [83]:
def clean_sleep_raw(path):
    df = pd.read_csv(path)
    df = df.drop_duplicates()
    df['Id'] = df['Id'].astype(str)
    df['SleepDay'] = pd.to_datetime(df['SleepDay'])
    df.columns = df.columns.str.lower()
    return df

In [84]:
sleep = clean_sleep_raw(sleep_path)
sleep.duplicated().sum()

/tmp/ipykernel_808/3803924614.py:5: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.



np.int64(0)

In [85]:
hourly_steps_raw = pd.read_csv(hourly_steps_path)
hourly_steps_raw.head()

,Id,ActivityHour,StepTotal
0,1503960366,4/12/2016 12:00:00 AM,373
1,1503960366,4/12/2016 1:00:00 AM,160
2,1503960366,4/12/2016 2:00:00 AM,151
3,1503960366,4/12/2016 3:00:00 AM,0
4,1503960366,4/12/2016 4:00:00 AM,0


In [86]:
def clean_hourly(path):
    df = pd.read_csv(path)
    df['Id'] = df['Id'].astype(str)
    df['ActivityHour'] = pd.to_datetime(df['ActivityHour'])
    df.columns = df.columns.str.lower()
    return df

In [87]:
hourly_steps = clean_hourly(hourly_steps_path)
hourly_calories = clean_hourly(hourly_calories_path)
hourly_intensities = clean_hourly(hourly_intensities_path)

/tmp/ipykernel_808/1786061669.py:4: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

/tmp/ipykernel_808/1786061669.py:4: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

/tmp/ipykernel_808/1786061669.py:4: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.



In [88]:
hourly = pd.merge(hourly_steps, hourly_calories, on=['id', 'activityhour'])
hourly = pd.merge(hourly, hourly_intensities, on=['id', 'activityhour'])
hourly.head()

,id,activityhour,steptotal,calories,totalintensity,averageintensity
0,1503960366,2016-04-12 00:00:00,373,81,20,0.333333
1,1503960366,2016-04-12 01:00:00,160,61,8,0.133333
2,1503960366,2016-04-12 02:00:00,151,59,7,0.116667
3,1503960366,2016-04-12 03:00:00,0,47,0,0.000000
4,1503960366,2016-04-12 04:00:00,0,48,0,0.000000


In [89]:
heartrate_raw = pd.read_csv(heartrate_path)
heartrate_raw.head()

,Id,Time,Value
0,2022484408,4/12/2016 7:21:00 AM,97
1,2022484408,4/12/2016 7:21:05 AM,102
2,2022484408,4/12/2016 7:21:10 AM,105
3,2022484408,4/12/2016 7:21:20 AM,103
4,2022484408,4/12/2016 7:21:25 AM,101


In [90]:
def clean_heartrate(path):
    df = pd.read_csv(path)
    df['Id'] = df['Id'].astype(str)
    df['Time'] = pd.to_datetime(df['Time'])
    df['hour'] = df['Time'].dt.floor('h')
    df.columns = df.columns.str.lower()
    df = df.groupby(['id', 'hour'])['value'].mean().reset_index()
    df.columns = ['id', 'hour', 'avg_heart_rate']
    return df

In [91]:
heartrate = clean_heartrate(heartrate_path)
heartrate.head()

,id,hour,avg_heart_rate
0,2022484408,2016-04-12 07:00:00,83.200000
1,2022484408,2016-04-12 08:00:00,68.562005
2,2022484408,2016-04-12 09:00:00,66.404700
3,2022484408,2016-04-12 10:00:00,106.716075
4,2022484408,2016-04-12 11:00:00,67.767157


## 2. Database Connection

Data is stored in **Supabase PostgreSQL** and queried directly into
Python using SQLAlchemy. This approach demonstrates the integration
of SQL and Python in a real-world data workflow.

In [92]:
!pip install sqlalchemy psycopg2-binary

In [93]:
from google.colab import userdata

DB_HOST     = userdata.get('DB_HOST')
DB_NAME     = userdata.get('DB_NAME')
DB_USER     = userdata.get('DB_USER')
DB_PASSWORD = userdata.get('DB_PASSWORD')
DB_PORT     = userdata.get('DB_PORT')

In [94]:
from sqlalchemy import create_engine

DATABASE_URL = f"postgresql://{DB_USER}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_NAME}"
engine = create_engine(DATABASE_URL)

print("連線成功！")

連線成功！


In [108]:
from sqlalchemy import text

with engine.begin() as conn:
    conn.execute(text("DROP VIEW IF EXISTS user_segments CASCADE"))
    conn.execute(text("DROP VIEW IF EXISTS user_sleep CASCADE"))
    conn.execute(text("DROP VIEW IF EXISTS user_time CASCADE"))

daily_activity.to_sql('daily_activity', engine, if_exists='replace', index=False)
print("daily_activity 完成！")

sleep.to_sql('sleep', engine, if_exists='replace', index=False)
print("sleep 完成！")

hourly.to_sql('hourly', engine, if_exists='replace', index=False)
print("hourly 完成！")

heartrate.to_sql('heartrate', engine, if_exists='replace', index=False)
print("heartrate 完成！")

daily_activity 完成！
sleep 完成！
hourly 完成！
heartrate 完成！


In [109]:
with engine.begin() as conn:
    conn.execute(text("""
        CREATE VIEW user_segments AS
        SELECT id, AVG(totalsteps) AS avg_steps,
          CASE
            WHEN AVG(totalsteps) <= 5000 THEN 'Sedentary'
            WHEN AVG(totalsteps) <= 7499 THEN 'Lightly Active'
            WHEN AVG(totalsteps) <= 9999 THEN 'Fairly Active'
            ELSE 'Very Active'
          END AS user_type
        FROM daily_activity
        GROUP BY id
    """))
    conn.execute(text("""
        CREATE VIEW user_time AS
        SELECT id, COUNT(DISTINCT activitydate) AS wear_time,
          CASE
            WHEN COUNT(DISTINCT activitydate) <= 15 THEN 'Low User'
            WHEN COUNT(DISTINCT activitydate) <= 24 THEN 'Moderate User'
            ELSE 'High User'
          END AS wearing_type
        FROM daily_activity
        GROUP BY id
    """))
    conn.execute(text("""
        CREATE VIEW user_sleep AS
        SELECT
          s.id, s.sleepday, s.totalminutesasleep, s.totaltimeinbed,
          s.totalminutesasleep::float / s.totaltimeinbed AS sleep_efficiency,
          d.totalsteps,
          LEAD(d.totalsteps) OVER (PARTITION BY d.id ORDER BY d.activitydate) AS next_day_steps,
          CASE WHEN s.totalminutesasleep::float / s.totaltimeinbed >= 0.9 THEN 'Good Sleep'
               ELSE 'Poor Sleep'
          END AS sleeping_type
        FROM sleep s
        JOIN daily_activity d ON s.id = d.id
        AND s.sleepday = d.activitydate
    """))

print("Views 建立完成！")

Views 建立完成！


In [97]:
print(engine)

Engine(postgresql://postgres.ftizpettqltxruirlmrl:***@aws-1-ap-southeast-1.pooler.supabase.com:5432/postgres)


## 3. Analysis

### 3.1 User Activity Segmentation

Users were segmented into 4 groups based on average daily steps:

| Segment | Average Daily Steps |
|---------|-------------------|
| Very Active | ≥ 10,000 |
| Fairly Active | 7,500 – 9,999 |
| Lightly Active | 5,000 – 7,499 |
| Sedentary | < 5,000 |

A PostgreSQL View was created to store the user segments analysis logic:

```sql
SELECT
    id,
    AVG(totalsteps) AS avg_steps,
    CASE
        WHEN AVG(totalsteps) <= 5000 THEN 'Sedentary'
        WHEN AVG(totalsteps) <= 7499 THEN 'Lightly Active'
        WHEN AVG(totalsteps) <= 9999 THEN 'Fairly Active'
        ELSE 'Very Active'
    END AS user_type
FROM daily_activity
GROUP BY id;
```

The distribution of users across segments was calculated as follows:

```sql
SELECT user_type, COUNT(*) AS user_count
FROM user_segments
GROUP BY user_type
ORDER BY user_count DESC;
```

**Finding:** Users are fairly evenly distributed across all four segments,
with Fairly Active and Lightly Active users each making up 27% of the sample.

In [98]:
user_segments = pd.read_sql("SELECT* FROM user_segments", engine)
user_segments.head()

,id,avg_steps,user_type
0,8253242879,6482.157895,Lightly Active
1,1644430081,7282.966667,Lightly Active
2,5577150313,8304.433333,Fairly Active
3,6962181067,9794.806452,Fairly Active
4,4057192912,3838.000000,Sedentary


In [99]:
user_type_count = user_segments['user_type'].value_counts().reset_index()
user_type_count.columns = ['user_type', 'count']

In [100]:
category_order = ['Very Active', 'Fairly Active', 'Lightly Active', 'Sedentary']
colors = ['#2C5F8A', '#4A86B8', '#7AAFD4', '#B8D4E8']

fig = px.pie(user_type_count,
             names='user_type',
             values='count',
             title='User Activity Segments',
             color_discrete_sequence=['#2C5F8A', '#4A86B8', '#7AAFD4', '#B8D4E8'],
             hole=0.5)

fig.update_layout(
    title_font_size=20,
)
fig.update_layout(
    title_font_size=20,
    annotations=[dict(
        text='33 users',
        x=0.5, y=0.5,
        font_size=16,
        showarrow=False
    )]
)
fig.show()

### 3.2 Average Calories by Activity Level

To understand the relationship between activity level and calorie burn,
average daily calories were calculated for each user segment.

```sql
SELECT
    user_segments.user_type,
    AVG(daily_activity.calories) AS avg_calories
FROM user_segments
JOIN daily_activity ON user_segments.id = daily_activity.id
GROUP BY user_segments.user_type
ORDER BY avg_calories DESC;
```

**Finding:** Very Active users burn approximately 537 more calories per day
than Sedentary users. Notably, the gap between Fairly Active and Very Active
is small (73 kcal), while Lightly Active and Fairly Active show a larger
jump (332 kcal).

In [101]:
calories_by_segment = pd.read_sql("""
    SELECT user_segments.user_type, AVG(daily_activity.calories) AS avg_calories
    FROM user_segments
    JOIN daily_activity ON user_segments.id = daily_activity.id
    GROUP BY user_segments.user_type
    ORDER BY avg_calories DESC
""", engine)
calories_by_segment.head()

,user_type,avg_calories
0,Very Active,2554.170616
1,Fairly Active,2481.660377
2,Lightly Active,2149.116000
3,Sedentary,2016.560748


In [102]:
category_order = ['Very Active', 'Fairly Active', 'Lightly Active', 'Sedentary']
colors = ['#2C5F8A', '#4A86B8', '#7AAFD4', '#B8D4E8']

fig = px.bar(calories_by_segment,
             x='user_type',
             y='avg_calories',
             title='Average Calories by Activity Level',
             category_orders={'user_type': category_order},
             color='user_type',
             color_discrete_sequence=colors,
             text='avg_calories')

fig.update_traces(texttemplate='%{text:.0f}', textposition='outside')
fig.update_layout(
    plot_bgcolor='white',
    paper_bgcolor='white',
    showlegend=False,
    xaxis_title='Activity Level',
    yaxis_title='Average Calories',
    title_font_size=20
)

fig.show()

### 3.3 Sleep Quality Analysis

Sleep quality was measured using **sleep efficiency** — the ratio of
actual sleep time to total time in bed — rather than sleep duration alone.

**Sleep Efficiency = Total Minutes Asleep / Total Time in Bed**

Users were categorized as:
- **Good Sleep:** Efficiency ≥ 90%
- **Poor Sleep:** Efficiency < 90%

A PostgreSQL View was created to store the sleep analysis logic:

```sql
SELECT
  s.id,
  s.totalminutesasleep::float / s.totaltimeinbed AS sleep_efficiency,
  d.totalsteps,
  LEAD(d.totalsteps) OVER (PARTITION BY d.id ORDER BY d.activitydate) AS next_day_steps,
  CASE WHEN s.totalminutesasleep::float / s.totaltimeinbed >= 0.9 THEN 'Good Sleep'
       ELSE 'Poor Sleep'
  END AS sleeping_type
FROM sleep s
JOIN daily_activity d ON s.id = d.id
AND s.sleepday = d.activitydate;
```

The following query was used to compare step counts between sleep quality groups:

```sql
SELECT sleeping_type,
       AVG(next_day_steps) AS avg_next_day,
       AVG(totalsteps) AS avg_total_steps
FROM user_sleep
GROUP BY sleeping_type;
```

Individual user sleep efficiency was also calculated:

```sql
SELECT id, sleeping_type,
       AVG(totalminutesasleep::float/totaltimeinbed) AS avg_sleeping_efficiency,
       AVG(next_day_steps) AS avg_next_day,
       AVG(totalsteps) AS avg_total_steps
FROM user_sleep
GROUP BY id, sleeping_type;
```

**Finding:** Counter-intuitively, Poor Sleep users show higher step counts
both on the same day (10,469 vs 8,359) and the next day (9,320 vs 8,361).
This suggests that higher activity levels may reduce sleep efficiency,
highlighting the importance of tracking both activity and recovery.

In [103]:
sleep_analysis = pd.read_sql("""
    SELECT sleeping_type,
           AVG(totalsteps) AS avg_total_steps,
           AVG(next_day_steps) AS avg_next_day_steps
    FROM user_sleep
    GROUP BY sleeping_type
""", engine)
sleep_analysis.head()

,sleeping_type,avg_total_steps,avg_next_day_steps
0,Good Sleep,8358.935976,8360.687097
1,Poor Sleep,9138.804878,9320.013158


In [104]:
sleep_melted = pd.melt(sleep_analysis,
                        id_vars='sleeping_type',
                        value_vars=['avg_total_steps', 'avg_next_day_steps'],
                        var_name='metric',
                        value_name='steps')

sleep_melted.head()

,sleeping_type,metric,steps
0,Good Sleep,avg_total_steps,8358.935976
1,Poor Sleep,avg_total_steps,9138.804878
2,Good Sleep,avg_next_day_steps,8360.687097
3,Poor Sleep,avg_next_day_steps,9320.013158


In [105]:
fig = px.bar(sleep_melted,
             x='sleeping_type',
             y='steps',
             color='metric',
             barmode='group',
             title='Steps by Sleep Quality',
             color_discrete_sequence=['#2C5F8A', '#7AAFD4'],
             text='steps')

fig.update_traces(texttemplate='%{text:.0f}', textposition='outside')
fig.update_layout(
    plot_bgcolor='white',
    paper_bgcolor='white',
    xaxis_title='Sleep Quality',
    yaxis_title='Average Steps',
    title_font_size=20,
    legend_title='Metric'
)
fig.update_layout(
    legend=dict(
        title='Metric',
        itemsizing='constant'
    )
)

newnames = {'avg_total_steps': 'Same Day Steps',
            'avg_next_day_steps': 'Next Day Steps'}

fig.for_each_trace(lambda t: t.update(name=newnames[t.name]))

fig.show()

### 3.4 Device Wearing Frequency

To identify user engagement with the device, users were categorized
based on the number of days they wore the device during the 31-day period.

| Segment | Days Worn |
|---------|-----------|
| High User | 25 – 31 days |
| Moderate User | 15 – 24 days |
| Low User | < 15 days |

A PostgreSQL View was created to categorize users by wearing frequency:

```sql
SELECT id, COUNT(DISTINCT activitydate) AS wear_time,
  CASE
    WHEN COUNT(DISTINCT activitydate) <= 15 THEN 'Low User'
    WHEN COUNT(DISTINCT activitydate) <= 24 THEN 'Moderate User'
    ELSE 'High User'
  END AS wearing_type
FROM daily_activity
GROUP BY id;
```

The distribution of users by wearing frequency:

```sql
SELECT wearing_type, COUNT(*) AS user_count
FROM user_time
GROUP BY wearing_type
ORDER BY user_count DESC;
```

**Finding:** 88% of users (29 out of 33) wore the device consistently
throughout the month, indicating high engagement with fitness tracking.
However, the remaining 12% represent an opportunity for Bellabeat to
improve user retention through targeted notifications and reminders.

In [106]:
wear_time = pd.read_sql("""
    SELECT wearing_type, COUNT(*) AS user_count
    FROM user_time
    GROUP BY wearing_type
    ORDER BY user_count DESC
""", engine)

In [107]:
category_order = ['High User', 'Moderate User', 'Low User' ]
colors = ['#4A86B8', '#7AAFD4', '#B8D4E8']

fig = px.bar(wear_time,
             x='user_count',
             y='wearing_type',
             orientation='h',
             title='User Wearing Frequency',
             category_orders={'wearing_type': category_order},
             color='wearing_type',
             color_discrete_sequence=colors,
             text='user_count')

fig.update_traces(textposition='outside')
fig.update_layout(
    plot_bgcolor='white',
    paper_bgcolor='white',
    showlegend=False,
    xaxis_title='Number of Users',
    yaxis_title='',
    title_font_size=20
)

fig.show()

## 4. Conclusion & Recommendations

### Key Findings

1. **Users are evenly distributed across activity levels** — 27% Fairly Active,
27% Lightly Active, 24% Sedentary, and 21% Very Active, suggesting Bellabeat
should design features for all activity levels, not just highly active users.

2. **Activity level strongly correlates with calorie burn** — Very Active users
burn 537 more calories daily than Sedentary users. The largest gap occurs
between Lightly Active and Fairly Active (332 kcal), suggesting this is the
most impactful transition to encourage.

3. **High activity may reduce sleep efficiency** — Counter-intuitively, users
with poor sleep efficiency show higher step counts. This highlights the need
for Bellabeat to promote balanced activity and recovery, not just movement.

4. **88% of users are highly engaged with their devices** — Consistent device
usage suggests strong habit formation. The remaining 12% represent a retention
opportunity.

### Recommendations for Bellabeat

1. **Introduce personalized activity goals** — Since users are spread across
all activity levels, Bellabeat's app should set adaptive goals based on each
user's current segment, nudging Sedentary users toward Lightly Active and
Lightly Active users toward Fairly Active.

2. **Promote the sleep-activity balance** — Market Bellabeat devices not just
as activity trackers, but as recovery monitors. Highlight the sleep efficiency
feature to differentiate from competitors.

3. **Re-engagement campaign for low-frequency users** — Target the 12% of
Moderate and Low users with push notifications, streak rewards, or reminders
to build consistent wearing habits.